mempool/src/tests/fuzzing.rs

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use super::super::{
+    core_mempool::CoreMempool,
+    shared_mempool::{tasks, types::SharedMempool},
+};
+use crate::core_mempool::TimelineState;
+
+use super::super::shared_mempool::peer_manager::PeerManager;
+use libra_config::config::NodeConfig;
+use libra_proptest_helpers::ValueGenerator;
+use libra_types::transaction::SignedTransaction;
+use rand::seq::SliceRandom;
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex, RwLock},
+};
+use storage_interface::mock::MockDbReader;
+use vm_validator::mocks::mock_vm_validator::MockVMValidator;
+
+#[test]
+fn test_mempool_fuzzer() {
+    let mut gen = ValueGenerator::new();
+    let data = generate_corpus(&mut gen);
+    fuzzer(&data);
+}
+
+/// Generates corpus for fuzzing mempool incoming transactions endpoint
+pub fn generate_corpus(gen: &mut ValueGenerator) -> Vec<u8> {
+    // generate 100 random SignedTransactions via proptest
+    // 100 is the max number of transactions in a submission batch
+    let mut txns = vec![];
+    for _ in 0..100 {
+        let txn = gen.generate(proptest::arbitrary::any::<SignedTransaction>());
+        txns.push(txn);
+    }
+    lcs::to_bytes(&txns)
+        .expect("[fuzzing] failed to convert mempool submitted transactions to bytes")
+}
+
+/// Fuzzes mempool incoming transaction endpoint with `data`
+pub fn fuzzer(data: &[u8]) {
+    // set up mock Shared Mempool
+    let config = NodeConfig::default();
+    let mock_db = MockDbReader;
+    let vm_validator = Arc::new(RwLock::new(MockVMValidator));
+    let smp = SharedMempool {
+        mempool: Arc::new(Mutex::new(CoreMempool::new(&config))),
+        config: config.mempool,
+        network_senders: HashMap::new(),
+        db: Arc::new(mock_db),
+        validator: vm_validator,
+        peer_manager: Arc::new(PeerManager::new(config.upstream)),
+        subscribers: vec![],
+    };
+
+    // generate fuzz data
+    let txns: Vec<SignedTransaction> = lcs::from_bytes(data)
+        .expect("[fuzzing] failed to convert bytes to transactions submitted to mempool");
+
+    // randomly select timeline state
+    let possible_timeline_states = vec![TimelineState::NotReady, TimelineState::NonQualified];
+    let timeline_state = possible_timeline_states
+        .choose(&mut rand::thread_rng())
```

Comment:
in general it's best to avoid RNGs in fuzzers because it is not deterministic (imagine trying to re-run it with an input you know will fail, but then doesn't because the other timeline_state gets called) and it slows down the run.

Here I would suggest instead using `prop_oneof![]` as part of a proptest strategy

Reply:
also what is this `choose()` function oO?

Reply:
maybe something like this would work?

```rust

fn mempool_something_strategy() -> impl Strategy<(Vec<SignedTransaction>, TimelineState)> {
  (
    proptest::collection::vec(any::<SignedTransaction>(), 100),
    prop_oneof![
      Just(TimelineState::NotReady), 
      Just(TimelineState::NonQualified),
    ],
  )
}
```